### DISCUSSION TITLE
Custom Protocols

### DISCUSSION THREAD

**Discussion by sairam4123** [godotengine/godot-proposals#4422] [General Discussions]
Let's say we want to have a protocol for assets, instead of doing `user://assets` everytime, we can do `assets://`
Just like how Scene Unique Nodes have been implemented, the file protocols will work the same way with some extended behaviour.
```
Protocols -- Paths
assets:// -->
            1. user://assets/
            2. res://assets/
music:// --> 
            1. assets://music/ # you can use previously defined protocol in the paths for the new protocol.
```
If file-paths in the host system are missing, then it will be created automatically.
If more than one file_path is provided, the resulting protocol will have both filepaths.
If two file paths in the protocol have same file name, the first file will be replaced by the second file.

It's also better to use this system if you have a big nested system of folders, this will be really helpful.

I hope I'm clear with what I'm trying to convey here.
Let me know if there's anything wrong with this approach.
I'll clearly state this:
```
1. load:
    - it's based on the order in which the file-paths are defined.
    - so if you declare it this way:
        -> assets://
              -> 1. res://assets/
              -> 2. user://assets/
        -> it will first check whether the file exists in the res folder, if not, then checks the user folder.
    - if you change up the order,
        -> it will check whether the file exists in the user folder, if not, then it loads from the res folder.
2. save:
    - writes to the first file path that is writable.
    - so if you declare the assets:// protocol like this:
        -> assets://
              -> 1. res://assets/
              -> 2. user://assets/
        -> it will first check res folder, if it's not writable, then it will save to user folder.
3. overwrite:
    - same as write but overwrites the first file with the same name.
    - so if you have two assets with same name in same protocol, writing to the file will replace the file in the writable file-path
```
For nested use of protocols, the protocols you can use are the protocols you have defined above.

It's like Folder Union, union of multiple folders as a single folder, and custom protocols create a proxy to the folder.

**Comment by Calinou** [comment]
I feel this proposal conflicts with Godot's [project organization](https://docs.godotengine.org/en/stable/tutorials/best_practices/project_organization.html) guidelines, which recommend grouping files by *purpose* rather than by *type*.

This also makes paths harder to copy-paste across projects, as custom protocols would be defined on a per-project basis with nonstandard paths in each project. This in turn can make refactoring and plugin management harder (what about conflicts between plugins?).

**Reply by sairam4123** [reply]
But what about Folder Unions? Would it be a good idea to have res:// and user:// be combined? Since it produces duplicate code since I have to open two different folders to get all the required assets since I have to support Modding.

**Reply by Calinou** [reply]
Folder unions can lead to conflicts when files use the same names under both actual filesystem locations.

Also, `user://` and `res://` serve different purposes. `user://` is *guaranteed* to be writable by the user in the exported project, while `res://` should *never* be assumed to be writable in an exported project (it generally isn't).

To improve modding support, I'd prefer the approach outlined in https://github.com/godotengine/godot-proposals/issues/2689 or one of its related proposals.

**Reply by sairam4123** [reply]
Welp, seems like Reduz has created a issue, so can this discussion be closed with a link to the issue?

**Reply by Calinou** [reply]
Indeed: https://github.com/godotengine/godot-proposals/issues/6307
